rust_src/src/frames.rs

```diff
@@ -305,4 +310,56 @@ pub fn frame_text_height(frame: LispObject) -> LispObject {
     LispObject::from(frame_or_selected(frame).text_height)
 }
 
+/// Return scroll bar width of FRAME in pixels.
+#[lisp_fn(min = "0", name = "scroll_bar_width")]
+pub fn frame_scroll_bar_width(frame: LispObject) -> LispObject {
+    let frame_ref = frame_or_selected(frame);
+    let width = match frame_ref.vertical_scroll_bar_type() {
+        vertical_scroll_bar_type::vertical_scroll_bar_left
+        | vertical_scroll_bar_type::vertical_scroll_bar_right => frame_ref.config_scroll_bar_width,
+        _ => 0,
+    };
+    LispObject::from(width)
+}
+
+/// Return scroll bar height of FRAME in pixels.
+#[lisp_fn(min = "0", name = "scroll_bar_height")]
+pub fn frame_scroll_bar_height(frame: LispObject) -> LispObject {
+    let frame_ref = frame_or_selected(frame);
+    let height = if frame_ref.horizontal_scroll_bars() {
+        frame_ref.config_scroll_bar_height
+    } else {
+        0
+    };
+    LispObject::from(height)
+}
+
+/// Return fringe width of FRAME in pixels.
+#[lisp_fn(min = "0", name = "fringe_width")]
+pub fn frame_fringe_width(frame: LispObject) -> LispObject {
+    let frame_ref = frame_or_selected(frame);
+    LispObject::from(frame_ref.total_fringe_width())
+}
+
+/// Return width of FRAME's internal border in pixels.
+#[lisp_fn(min = "0")]
+pub fn frame_internal_border_width(frame: LispObject) -> LispObject {
+    let frame_ref = frame_or_selected(frame);
+    LispObject::from(frame_ref.internal_border_width())
+}
+
+/// Return width (in pixels) of vertical window dividers on FRAME.
+#[lisp_fn(min = "0", name = "right_divider_width")]
+pub fn frame_right_divider_width(frame: LispObject) -> LispObject {
+    let frame_ref = frame_or_selected(frame);
+    LispObject::from(frame_ref.right_divider_width)
+}
+
+/// Return width (in pixels) of horizontal window dividers on FRAME.
+#[lisp_fn(min = "0", name = "bottom_divider_width")]
```

Comment:
The tests are failing in this PR, and I think it's because you've set the name. Remacs isn't getting a function called `frame-bottom-dviider-width`. I think this should be the other way round:

```
#lispfn(name="frame-bottom-divider-width")
pub fn bottom_divider_width(...)
```

Reply:
Why is `name` being used at all here? Is there a conflict somewhere?

Reply:
```
src/frame.c
3171 DEFUN ("frame-bottom-divider-width", Fbottom_divider_width, Sbottom_divider_width, 0, 1, 0,
3175   return make_number (FRAME_BOTTOM_DIVIDER_WIDTH (decode_any_frame (frame)));
3312   {"bottom-divider-width",	SYMBOL_INDEX (Qbottom_divider_width)},
3855   store_in_alist (alistptr, Qbottom_divider_width,
3856 		  make_number (FRAME_BOTTOM_DIVIDER_WIDTH (f)));
4235 x_set_bottom_divider_width (struct frame *f, Lisp_Object arg, Lisp_Object oldval)
4237   int old = FRAME_BOTTOM_DIVIDER_WIDTH (f);
4242       f->bottom_divider_width = new;
4243       adjust_frame_size (f, -1, -1, 4, 0, Qbottom_divider_width);
5444   DEFSYM (Qbottom_divider_width, "bottom-divider-width");
5847   defsubr (&Sbottom_divider_width);
```

It's consistent with the C code the Rust is replacing (the C code had different names), but I don't see any obvious conflict. We could probably remove it as far as I can see.

Reply:
Oh I missed this. I take a look at it again.